File: src/main/scala/com/ozancicek/artan/ml/filter/FilterParams.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ozancicek.artan.ml.filter

import org.apache.spark.ml.linalg.{DenseVector, DenseMatrix, Vector, Matrix}
import org.apache.spark.ml.param._


/**
 * Param for initial state vector of a filter.
 */
private[artan] trait HasInitialState extends Params {

  def stateSize: Int

  final val initialState: Param[Vector] = new Param[Vector](
    this,
    "initialState",
    "Initial value of the state vector",
    (in: Vector) => in.size == stateSize)

  setDefault(initialState, new DenseVector(Array.fill(stateSize) {0.0}))

  final def getInitialState: Vector = $(initialState)

}


/**
 * Param for initial covariance of the state
 */
private[artan] trait HasInitialCovariance extends Params {

  def stateSize: Int

  final val initialCovariance: Param[Matrix] = new Param[Matrix](
    this,
    "initialCovariance",
    "Initial covariance matrix",
    (in: Matrix) => (in.numRows == stateSize) & (in.numCols == stateSize))

  setDefault(initialCovariance, DenseMatrix.eye(stateSize))

  final def getInitialCovariance: Matrix = $(initialCovariance)

}


/**
 * Param for process model matrix
 */
private[artan] trait HasProcessModel extends Params {

  def stateSize: Int

  final val processModel: Param[Matrix] = new Param[Matrix](
    this,
    "processModel",
    "Process model matrix, transitions the state to the next state when applied",
    (in: Matrix) => (in.numRows == stateSize) & (in.numCols == stateSize))

  setDefault(processModel, DenseMatrix.eye(stateSize))

  final def getProcessModel: Matrix = $(processModel)
}


/**
 * Param for fading factor
 */
private[artan] trait HasFadingFactor extends Params {

  final val fadingFactor: DoubleParam = new DoubleParam(
    this,
    "fadingFactor",
    "Factor controlling the weight of older measurements. With larger factor, more weights will" +
    "be given to recent measurements. Typically, should be really close to 1",
    ParamValidators.gtEq(1.0))

  setDefault(fadingFactor, 1.0)

  final def getFadingFactor: Double = $(fadingFactor)
}


/**
 * Param for measurement model matrix
 */
private[artan] trait HasMeasurementModel extends Params {

  def stateSize: Int
  def measurementSize: Int

  final val measurementModel: Param[Matrix] = new Param[Matrix](
    this,
    "measurementModel",
    "Measurement matrix, when multiplied with the state it should give the measurement vector",
    (in: Matrix) => (in.numRows == measurementSize) & (in.numCols == stateSize))

  setDefault(
    measurementModel,
    new DenseMatrix(
      measurementSize,
      stateSize,
      1.0 +: Array.fill(stateSize * measurementSize - 1) {0.0}))

  final def getMeasurementModel: Matrix = $(measurementModel)

}


/**
 * Param for process noise matrix
 */
private[artan] trait HasProcessNoise extends Params {

  def stateSize: Int

  final val processNoise: Param[Matrix] = new Param[Matrix](
    this,
    "processNoise",
    "Process noise matrix")

  setDefault(processNoise, DenseMatrix.eye(stateSize))

  final def getProcessNoise: Matrix = $(processNoise)
}


/**
 * Param for measurement noise matrix
 */
private[artan] trait HasMeasurementNoise extends Params {

  def measurementSize: Int

  final val measurementNoise: Param[Matrix] = new Param[Matrix](
    this,
    "measurementNoise",
    "Measurement noise matrix")

  setDefault(measurementNoise, DenseMatrix.eye(measurementSize))

  final def getMeasurementNoise: Matrix = $(measurementNoise)
}


/**
 * Param for process function, typically for nonlinear state transition processes
 */
private[artan] trait HasProcessFunction extends Params {

  final val processFunction: Param[(Vector, Matrix) => Vector] = new Param[(Vector, Matrix) => Vector](
    this,
    "processFunction",
    "Process function for nonlinear state transition. Input args to the function are state vector" +
    "and process model matrix. It should output the next state vector"
  )

  final def getProcessFunctionOpt: Option[(Vector, Matrix) => Vector] = get(processFunction)
}


/**
 * Param for process state jacobian for nonlinear state transition process
 */
private[artan] trait HasProcessStateJacobian extends Params {

  final val processStateJacobian: Param[(Vector, Matrix) => Matrix] = new Param[(Vector, Matrix) => Matrix](
    this,
    "processStateJacobian",
    "Process state jacobian function. Input args to the function are state vector" +
    "and process model matrix. It should output the jacobian matrix at input state"
  )

  final def getProcessStateJacobianOpt: Option[(Vector, Matrix) => Matrix] = get(processStateJacobian)
}


/**
 * Param for process noise jacobian for non-additive process noise
 */
private[artan] trait HasProcessNoiseJacobian extends Params {

  final val processNoiseJacobian: Param[(Vector, Matrix) => Matrix] = new Param[(Vector, Matrix) => Matrix](
    this,
    "processNoiseJacobian",
    "Process noise jacobian function. Input args to the function are the current state and process noise," +
    "it should output the process noise jacobian at input state. The output dimensions of the noise jacobian matrix" +
    "should be (n_state, n_noise), where n_state is the length of state vector and n_noise is the dimension of" +
    "the square processNoise matrix. Namely, let Q be an (n_noise, n_noise) matrix for process noise, and" +
    "let H_k(x_k) be the noise jacobian at state x_k. The result of H_k(x_k) * Q * H_k(x_k).T should have" +
    "(n_state, n_state) dimensions"
  )

  final def getProcessNoiseJacobianOpt: Option[(Vector, Matrix) => Matrix] = get(processNoiseJacobian)
}


/**
 * Param for measurement function, typically for nonlinear measurement equations.
 */
private[artan] trait HasMeasurementFunction extends Params {

  final val measurementFunction: Param[(Vector, Matrix) => Vector] = new Param[(Vector, Matrix) => Vector](
    this,
    "measurementFunction",
    "Measurement function for nonlinear measurement equations. Input args to the function are current" +
    "state and measurement model matrix. It should output the measurement corresponding to the state"
  )

  final def getMeasurementFunctionOpt: Option[(Vector, Matrix) => Vector] = get(measurementFunction)
}


/**
 * Param for measurement jacobian for nonlinear measurement equations
 */
private[artan] trait HasMeasurementStateJacobian extends Params {

  final val measurementStateJacobian: Param[(Vector, Matrix) => Matrix] = new Param[(Vector, Matrix) => Matrix](
    this,
    "measurementStateJacobian",
    "Measurement state jacobian function. Input args to the function are current state and measurement model" +
    "matrix. It should output the jacobian of measurement equation at input state"
  )

  final def getMeasurementStateJacobianOpt: Option[(Vector, Matrix) => Matrix] = get(measurementStateJacobian)
}


/**
 * Param for measurement noise jacobian for non-additive process noise
 */
private[artan] trait HasMeasurementNoiseJacobian extends Params {

  final val measurementNoiseJacobian: Param[(Vector, Matrix) => Matrix] = new Param[(Vector, Matrix) => Matrix](
    this,
    "measurementNoiseJacobian",
    "Measurement noise jacobian function. Input args to the function are the current state and" +
      "measurement noise, it should output the measurement noise jacobian at input state. " +
      "The output dimensions of the measurement noise jacobian matrix should be (n_measurement, n_noise)," +
      "where n_measurement is the length of measurement vector and n_noise is the dimension of" +
      "the square measurementNoise matrix. Namely, let Q be an (n_noise, n_noise) matrix for process noise, and" +
      "let H_k(x_k) be the noise jacobian at state x_k. The result of H_k(x_k) * Q * H_k(x_k).T should have" +
      "(n_measurement, n_measurement) dimensions"
  )

  final def getMeasurementNoiseJacobianOpt: Option[(Vector, Matrix) => Matrix] = get(measurementNoiseJacobian)
}

/**
 * Param for initial state column
 */
private[artan] trait HasInitialStateCol extends Params {

  final val initialStateCol: Param[String] = new Param[String](
    this,
    "initialStateCol",
    "Column name for initial state vector")

  final def getInitialStateCol: String = $(initialStateCol)
}


/**
 * Param for initial state covariance column
 */
private[artan] trait HasInitialCovarianceCol extends Params {

  final val initialCovarianceCol: Param[String] = new Param[String](
    this,
    "initialCovarianceCol",
    "Column name for initial covariance matrix")

  final def getInitialCovarianceCol: String = $(initialCovarianceCol)
}

/**
 * Param for measurement column
 */
private[artan] trait HasMeasurementCol extends Params {

  final val measurementCol: Param[String] = new Param[String](
    this,
    "measurementCol",
    "Column name for measurement vector. Missing measurements are allowed with nulls in the data")

  setDefault(measurementCol, "measurement")

  final def getMeasurementCol: String = $(measurementCol)
}


/**
 * Param for measurement model column
 */
private[artan] trait HasMeasurementModelCol extends Params {

  final val measurementModelCol: Param[String] = new Param[String](
    this,
    "measurementModelCol",
    "Column name for specifying measurement model from input DataFrame rather than" +
    "a constant measurement model for all filters")

  final def getMeasurementModelCol: String = $(measurementModelCol)
}


/**
 * Param for measurement noise column
 */
private[artan] trait HasMeasurementNoiseCol extends Params {

  final val measurementNoiseCol: Param[String] = new Param[String](
    this,
    "measurementNoiseCol",
    "Column name for specifying measurement noise from input DataFrame rather than" +
    "a constant measurement noise for all filters")

  final def getMeasurementNoiseCol: String = $(measurementNoiseCol)
}


/**
 * Param for process model column
 */
private[artan] trait HasProcessModelCol extends Params {

  final val processModelCol: Param[String] = new Param[String](
    this,
    "processModelCol",
    "Column name for specifying process model from input DataFrame rather than" +
    "a constant process model for all filters")

  final def getProcessModelCol: String = $(processModelCol)
}

/**
 * Param for process noise column
 */
private[artan] trait HasProcessNoiseCol extends Params {

  final val processNoiseCol: Param[String] = new Param[String](
    this,
    "processNoiseCol",
    "Column name for specifying process noise matrix from input DataFrame rather than" +
    "a constant process noise for all filters")

  final def getProcessNoiseCol: String = $(processNoiseCol)
}


/**
 * Param for control column
 */
private[artan] trait HasControlCol extends Params {

  final val controlCol: Param[String] = new Param[String](
    this,
    "controlCol",
    "Column name for specifying control vector")

  final def getControlCol: String = $(controlCol)
}


/**
 * Param for control function column
 */
private[artan] trait HasControlFunctionCol extends Params {

  final val controlFunctionCol: Param[String] = new Param[String](
    this,
    "controlFunctionCol",
    "Column name for specifying control matrix")

  final def getControlFunctionCol: String = $(controlFunctionCol)
}


/**
 * Param for enabling mahalanobis calculation
 */
private[artan] trait HasCalculateMahalanobis extends Params {

  final val calculateMahalanobis: BooleanParam = new BooleanParam(
    this,
    "calculateMahalanobis",
    "When true, mahalanobis distance of residual will be calculated & added to output DataFrame." +
    "Default is false.")

  setDefault(calculateMahalanobis, false)

  final def getCalculateMahalanobis: Boolean = $(calculateMahalanobis)
}


/**
 * Param for enabling loglikelihood calculation
 */
private[artan] trait HasCalculateLoglikelihood extends Params {

  final val calculateLoglikelihood: BooleanParam = new BooleanParam(
    this,
    "calculateLoglikelihood",
    "When true, loglikelihood of residual will be calculated & added to output DataFrame. Default is false")

  setDefault(calculateLoglikelihood, false)

  final def getCalculateLoglikelihood: Boolean = $(calculateLoglikelihood)
}

File: python/artan/filter/filter_params.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at

#  http://www.apache.org/licenses/LICENSE-2.0

#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

from pyspark.ml.param import Params, Param, TypeConverters


class HasInitialState(Params):
    """
    Mixin for initial state vector.
    """

    initialState = Param(
        Params._dummy(),
        "initialState", "Initial state vector", typeConverter=TypeConverters.toVector)

    def __init__(self):
        super(HasInitialState, self).__init__()

    def getInitialState(self):
        """
        Gets the value of initial state vector or its default value.
        """
        return self.getOrDefault(self.initialState)


class HasInitialStateCol(Params):
    """
    Mixin for param for initial state column.
    """

    initialStateCol = Param(
        Params._dummy(),
        "initialStateCol",
        "Column name for initial state vector.",
        typeConverter=TypeConverters.toString)

    def __init__(self):
        super(HasInitialStateCol, self).__init__()

    def getInitialStateCol(self):
        """
        Gets the value of initial state column or its default value.
        """
        return self.getOrDefault(self.initialStateCol)


class HasInitialCovariance(Params):
    """
    Mixin for param initial covariance matrix.
    """

    initialCovariance = Param(
        Params._dummy(),
        "initialCovariance",
        "Initial covariance matrix",
        typeConverter=TypeConverters.toMatrix)

    def __init__(self):
        super(HasInitialCovariance, self).__init__()

    def getInitialCovariance(self):
        """
        Gets the value of initial covariance matrix or its default value.
        """
        return self.getOrDefault(self.initialCovariance)


class HasInitialCovarianceCol(Params):
    """
    Mixin for param for initial covariance column.
    """

    initialCovarianceCol = Param(
        Params._dummy(),
        "initialCovarianceCol",
        "Column name for initial covariance vector.",
        typeConverter=TypeConverters.toString)

    def __init__(self):
        super(HasInitialCovarianceCol, self).__init__()

    def getInitialCovarianceCol(self):
        """
        Gets the value of initial covariance column or its default value.
        """
        return self.getOrDefault(self.initialCovarianceCol)


class HasProcessModel(Params):
    """
    Mixin for param process model matrix.
    """

    processModel = Param(
        Params._dummy(),
        "processModel",
        "Process model matrix, transitions the state to the next state when applied",
        typeConverter=TypeConverters.toMatrix)

    def __init__(self):
        super(HasProcessModel, self).__init__()

    def getProcessModel(self):
        """
        Gets the value of process model matrix or its default value.
        """
        return self.getOrDefault(self.processModel)


class HasFadingFactor(Params):
    """
    Mixin for param fading factor.
    """

    fadingFactor = Param(
        Params._dummy(),
        "fadingFactor",
        "Factor controlling the weight of older measurements. With larger factor, more weights will" +
        "be given to recent measurements. Typically, should be really close to 1",
        typeConverter=TypeConverters.toFloat)

    def __init__(self):
        super(HasFadingFactor, self).__init__()

    def getFadingFactor(self):
        """
        Gets the value of fading factor or its default value.
        """
        return self.getOrDefault(self.fadingFactor)


class HasMeasurementModel(Params):
    """
    Mixin for param measurement model matrix.
    """

    measurementModel = Param(
        Params._dummy(),
        "measurementModel",
        "Measurement matrix, when multiplied with the state it should give the measurement vector",
        typeConverter=TypeConverters.toMatrix)

    def __init__(self):
        super(HasMeasurementModel, self).__init__()

    def getMeasurementModel(self):
        """
        Gets the value of measurement model matrix or its default value.
        """
        return self.getOrDefault(self.measurementModel)


class HasProcessNoise(Params):
    """
    Mixin for param process noise matrix.
    """

    processNoise = Param(
        Params._dummy(),
        "processNoise",
        "Process noise matrix",
        typeConverter=TypeConverters.toMatrix)

    def __init__(self):
        super(HasProcessNoise, self).__init__()

    def getProcessNoise(self):
        """
        Gets the value of process noise matrix or its default value.
        """
        return self.getOrDefault(self.processNoise)


class HasMeasurementNoise(Params):
    """
    Mixin for param measurement noise matrix.
    """

    measurementNoise = Param(
        Params._dummy(),
        "measurementNoise",
        "Measurement noise matrix",
        typeConverter=TypeConverters.toMatrix)

    def __init__(self):
        super(HasMeasurementNoise, self).__init__()

    def getMeasurementNoise(self):
        """
        Gets the value of measurement noise matrix or its default value.
        """
        return self.getOrDefault(self.measurementNoise)


class HasMeasurementCol(Params):
    """
    Mixin for param for measurement column.
    """

    measurementCol = Param(
        Params._dummy(),
        "measurementCol",
        "Column name for measurement vector. Missing measurements are allowed with nulls in the data",
        typeConverter=TypeConverters.toString)

    def __init__(self):
        super(HasMeasurementCol, self).__init__()

    def getMeasurementCol(self):
        """
        Gets the value of measurement column or its default value.
        """
        return self.getOrDefault(self.measurementCol)


class HasMeasurementModelCol(Params):
    """
    Mixin for param for measurement model column.
    """

    measurementModelCol = Param(
        Params._dummy(),
        "measurementModelCol",
        "Column name for specifying measurement model from input DataFrame rather than" +
        "a constant measurement model for all filters",
        typeConverter=TypeConverters.toString)

    def __init__(self):
        super(HasMeasurementModelCol, self).__init__()

    def getMeasurementModelCol(self):
        """
        Gets the value of measurement model column or its default value.
        """
        return self.getOrDefault(self.measurementModelCol)


class HasMeasurementNoiseCol(Params):
    """
    Mixin for param for measurement noise column.
    """

    measurementNoiseCol = Param(
        Params._dummy(),
        "measurementNoiseCol",
        "Column name for specifying measurement noise from input DataFrame rather than" +
        "a constant measurement noise for all filters",
        typeConverter=TypeConverters.toString)

    def __init__(self):
        super(HasMeasurementNoiseCol, self).__init__()

    def getMeasurementNoiseCol(self):
        """
        Gets the value of measurement noise column or its default value.
        """
        return self.getOrDefault(self.measurementNoiseCol)


class HasProcessModelCol(Params):
    """
    Mixin for param for process model column.
    """

    processModelCol = Param(
        Params._dummy(),
        "processModelCol",
        "Column name for specifying process model from input DataFrame rather than" +
        "a constant measurement model for all filters",
        typeConverter=TypeConverters.toString)

    def __init__(self):
        super(HasProcessModelCol, self).__init__()

    def getProcessModelCol(self):
        """
        Gets the value of process model column or its default value.
        """
        return self.getOrDefault(self.processModelCol)


class HasProcessNoiseCol(Params):
    """
    Mixin for param for process noise column.
    """

    processNoiseCol = Param(
        Params._dummy(),
        "processNoiseCol",
        "Column name for specifying process noise from input DataFrame rather than" +
        "a constant measurement noise for all filters",
        typeConverter=TypeConverters.toString)

    def __init__(self):
        super(HasProcessNoiseCol, self).__init__()

    def getProcessNoiseCol(self):
        """
        Gets the value of process noise column or its default value.
        """
        return self.getOrDefault(self.processNoiseCol)


class HasControlCol(Params):
    """
    Mixin for param for control column.
    """

    controlCol = Param(
        Params._dummy(),
        "controlCol",
        "Column name for specifying control vector",
        typeConverter=TypeConverters.toString)

    def __init__(self):
        super(HasControlCol, self).__init__()

    def getControlCol(self):
        """
        Gets the value of control column or its default value.
        """
        return self.getOrDefault(self.controlCol)


class HasControlFunctionCol(Params):
    """
    Mixin for param for control function column.
    """

    controlFunctionCol = Param(
        Params._dummy(),
        "controlFunctionCol",
        "Column name for specifying control matrix",
        typeConverter=TypeConverters.toString)

    def __init__(self):
        super(HasControlFunctionCol, self).__init__()

    def getControlFunctionCol(self):
        """
        Gets the value of control function column or its default value.
        """
        return self.getOrDefault(self.controlFunctionCol)


class HasCalculateMahalanobis(Params):
    """
    Mixin for param for enabling mahalanobis calculation.
    """

    calculateMahalanobis = Param(
        Params._dummy(),
        "calculateMahalanobis",
        "When true, mahalanobis distance of residual will be calculated & added to output DataFrame." +
        "Default is false.",
        typeConverter=TypeConverters.toBoolean)

    def __init__(self):
        super(HasCalculateMahalanobis, self).__init__()

    def getCalculateMahalanobis(self):
        """
        Gets the value of mahalanobis calcuation flag.
        """
        return self.getOrDefault(self.calculateMahalanobis)


class HasCalculateLoglikelihood(Params):
    """
    Mixin for param for enabling loglikelihood calculation.
    """

    calculateLoglikelihood = Param(
        Params._dummy(),
        "calculateLoglikelihood",
        "When true, loglikelihood of residual will be calculated & added to output DataFrame. Default is false",
        typeConverter=TypeConverters.toBoolean)

    def __init__(self):
        super(HasCalculateLoglikelihood, self).__init__()

    def getCalculateLoglikelihood(self):
        """
        Gets the value of loglikelihood calculation flag.
        """
        return self.getOrDefault(self.calculateLoglikelihood)


class HasOutputSystemMatrices(Params):
    """
    Mixin for param for enabling the output of system matrices along with the state.
    """

    outputSystemMatrices = Param(
        Params._dummy(),
        "outputSystemMatrices",
        "When true, the system matrices will be added to output DataFrame. Default is false",
        typeConverter=TypeConverters.toBoolean)

    def __init__(self):
        super(HasOutputSystemMatrices, self).__init__()

    def getOutputSystemMatrices(self):
        """
        Gets the value of loglikelihood calculation flag.
        """
        return self.getOrDefault(self.outputSystemMatrices)


class HasCalculateSlidingLikelihood(Params):
    """
    Mixin param for enabling sliding likelihood calculation
    """
    calculateSlidingLikelihood = Param(
        Params._dummy(),
        "calculateSlidingLikelihood",
        "When true, sliding likelihood sum of residual will be calculated & added to output DataFrame. Default is false",
        typeConverter=TypeConverters.toBoolean)

    def __init__(self):
        super(HasCalculateSlidingLikelihood, self).__init__()

    def getCalculateSlidingLikelihood(self):
        """
        Gets the value of sliding likelihood calculation flag
        """
        return self.getOrDefault(self.calculateSlidingLikelihood)


class HasSlidingLikelihoodWindow(Params):
    """
    Mixin param for sliding likelihood window duration
    """
    slidingLikelihoodWindow = Param(
        Params._dummy(),
        "slidingLikelihoodWindow",
        "Number of consecutive measurements to include in the total likelihood calculation",
        typeConverter=TypeConverters.toInt)

    def __init__(self):
        super(HasSlidingLikelihoodWindow, self).__init__()

    def getSlidingLikelihoodWindow(self):
        """
        Gets the value of sliding likelihood window
        """
        return self.getOrDefault(self.slidingLikelihoodWindow)


class HasMultipleModelMeasurementWindowDuration(Params):

    multipleModelMeasurementWindowDuration = Param(
        Params._dummy(),
        "multipleModelMeasurementWindowDuration",
        "Window duration for grouping measurements in same window for MMAE filter aggregation",
        typeConverter=TypeConverters.toString)

    def __init__(self):
        super(HasMultipleModelMeasurementWindowDuration, self).__init__()

    def getMultipleModelMeasurementWindowDuration(self):
        """
        Gets the value of mmae measureent window duration
        """
        return self.getOrDefault(self.multipleModelMeasurementWindowDuration)


class HasMultipleModelAdaptiveEstimationEnabled(Params):

    multipleModelAdaptiveEstimationEnabled = Param(
        Params._dummy(),
        "multipleModelAdaptiveEstimationEnabled",
        "Flag for enabling  Multiple Model Adaptive Estimation (MMAE) output mode. When enabled," + "" +
        "MMAE mode outputs a single state estimate from the output of all kalman states of the transformer." +
        "States are weighted based on their sliding likelihood",
        typeConverter=TypeConverters.toBoolean)

    def __init__(self):
        super(HasMultipleModelAdaptiveEstimationEnabled, self).__init__()

    def getMultipleModelAdaptiveEstimationEnabled(self):
        """
        Gets the value of MMAE output mode flag
        """
        return self.getOrDefault(self.multipleModelAdaptiveEstimationEnabled)


class KalmanFilterParams(HasInitialState, HasInitialCovariance, HasInitialStateCol,
                         HasInitialCovarianceCol, HasProcessModel, HasFadingFactor, HasMeasurementModel,
                         HasMeasurementNoise, HasProcessNoise, HasMeasurementCol,
                         HasMeasurementModelCol, HasMeasurementNoiseCol, HasProcessModelCol,
                         HasProcessNoiseCol, HasControlCol, HasControlFunctionCol,
                         HasCalculateMahalanobis, HasCalculateLoglikelihood,
                         HasOutputSystemMatrices, HasCalculateSlidingLikelihood,
                         HasSlidingLikelihoodWindow):
    """
    Mixin for kalman filter parameters
    """
    def setInitialState(self, value):
        """
        Set the initial state vector with size (stateSize).

        It will be applied to all states. If the state timeouts and starts receiving
        measurements after timeout, it will again start from this initial state vector. Default is zero.

        Note that if this parameter is set through here, it will result in same initial state for all filters.
        For different initial states across filters, set the dataframe column for corresponding to initial state
        with setInitialStateCol.

        :param value: pyspark.ml.linalg.Vector with size (stateSize)
        :return: KalmanFilter
        """
        return self._set(initialState=value)

    def setInitialStateCol(self, value):
        """
        Set the column corresponding to initial state vector. Overrides setInitialState setting.

        :param value: String
        :return: KalmanFilter
        """
        return self._set(initialStateCol=value)

    def setInitialCovariance(self, value):
        """
        Set the initial covariance matrix with dimensions (stateSize, stateSize)

        It will be applied to all states. If the state timeouts and starts receiving
        measurements after timeout, it will again start from this initial covariance vector. Default is identity matrix.
        :param value: pyspark.ml.linalg.Matrix with dimensions (stateSize, stateSize)
        :return: KalmanFilter
        """
        return self._set(initialCovariance=value)

    def setInitialCovarianceCol(self, value):
        """
        Set the column corresponding to initial covariance matrix. Overrides setInitialCovariance setting.

        :param value: String
        :return: KalmanFilter
        """
        return self._set(initialCovarianceCol=value)

    def setFadingFactor(self, value):
        """
        Fading factor for giving more weights to more recent measurements. If needed, it should be greater than one.
        Typically set around 1.01 ~ 1.05. Default is 1.0, which will result in equally weighted measurements.

        :param value: Float >= 1.0
        :return: KalmanFilter
        """
        return self._set(fadingFactor=value)

    def setProcessModel(self, value):
        """
        Set default value for process model matrix with dimensions (stateSize, stateSize) which governs
        state transition.

        Note that if this parameter is set through here, it will result in same process model for all filters &
        measurements. For different process models across filters or measurements, set a dataframe column for process
        model from setProcessModelCol.

        Default is identity matrix.

        :param value: pyspark.ml.linalg.Matrix with dimensions (stateSize, stateSize)
        :return: KalmanFilter
        """
        return self._set(processModel=value)

    def setProcessNoise(self, value):
        """
        Set default value for process noise matrix with dimensions (stateSize, stateSize).

        Note that if this parameter is set through here, it will result in same process noise for all filters &
        measurements. For different process noise values across filters or measurements, set a dataframe column
        for process noise from setProcessNoiseCol.

        Default is identity matrix.

        :param value: pyspark.ml.linalg.Matrix with dimensions (stateSize, StateSize)
        :return: KalmanFilter
        """
        return self._set(processNoise=value)

    def setMeasurementModel(self, value):
        """
        Set default value for measurement model matrix with dimensions (stateSize, measurementSize)
        which maps states to measurement.

        Note that if this parameter is set through here, it will result in same measurement model for all filters &
        measurements. For different measurement models across filters or measurements, set a dataframe column for
        measurement model from setMeasurementModelCol.

        Default value maps the first state value to measurements.

        :param value: pyspark.ml.linalg.Matrix with dimensions (stateSize, measurementSize)
        :return: KalmanFilter
        """
        return self._set(measurementModel=value)

    def setMeasurementNoise(self, value):
        """
        Set default value for measurement noise matrix with dimensions (measurementSize, measurementSize).

        Note that if this parameter is set through here, it will result in same measurement noise for all filters &
        measurements. For different measurement noise values across filters or measurements,
        set a dataframe column for measurement noise from setMeasurementNoiseCol.

        Default is identity matrix.

        :param value: pyspark.ml.linalg.Matrix with dimensions (measurementSize, measurementSize)
        :return: KalmanFilter
        """
        return self._set(measurementNoise=value)

    def setMeasurementCol(self, value):
        """
        Set the column corresponding to measurements.

        The vectors in the column should be of size (measurementSize). null values are allowed,
        which will result in only state prediction step.

        :param value: pyspark.ml.linalg.Vector with size measurementSize
        :return: KalmanFilter
        """
        return self._set(measurementCol=value)

    def setMeasurementModelCol(self, value):
        """
        Set the column for input measurement model matrices

        Measurement model matrices should have dimensions (stateSize, measurementSize)

        :param value: String
        :return: KalmanFilter
        """
        return self._set(measurementModelCol=value)

    def setProcessModelCol(self, value):
        """
        Set the column for input process model matrices.

        Process model matrices should have dimensions (stateSize, stateSize)

        :param value: String
        :return: KalmanFilter
        """
        return self._set(processModelCol=value)

    def setProcessNoiseCol(self, value):
        """
        Set the column for input process noise matrices.

        Process noise matrices should have dimensions (stateSize, stateSize)

        :param value: String
        :return: KalmanFilter
        """
        return self._set(processNoiseCol=value)

    def setMeasurementNoiseCol(self, value):
        """
        Set the column for input measurement noise matrices.

        Measurement noise matrices should have dimensions (measurementSize, measurementSize)

        :param value: String
        :return: KalmanFilter
        """
        return self._set(measurementNoiseCol=value)

    def setControlCol(self, value):
        """
        Set the column for input control vectors.

        Control vectors should have compatible size with control function (controlVectorSize). The product of
        control matrix & vector should produce a vector with stateSize. null values are allowed,
        which will result in state transition without control input.

        :param value: String
        :return: KalmanFilter
        """
        return self._set(controlCol=value)

    def setControlFunctionCol(self, value):
        """
        Set the column for input control matrices.

        Control matrices should have dimensions (stateSize, controlVectorSize). null values are allowed, which will
        result in state transition without control input

        :param value: String
        :return: KalmanFilter
        """
        return self._set(controlFunctionCol=value)

    def setCalculateLogLikelihood(self):
        """
        Optionally calculate loglikelihood of each measurement & add it to output dataframe. Loglikelihood is calculated
        from residual vector & residual covariance matrix.

        Not enabled by default.

        :return: KalmanFilter
        """
        return self._set(calculateLoglikelihood=True)

    def setCalculateMahalanobis(self):
        """
        Optionally calculate mahalanobis distance metric of each measurement & add it to output dataframe.
        Mahalanobis distance is calculated from residual vector & residual covariance matrix.

        Not enabled by default.

        :return: KalmanFilter
        """
        return self._set(calculateMahalanobis=True)

    def setOutputSystemMatrices(self):
        """
        Optionally add system matrices to output dataframe returned by the transformer.

        Default is false

        :return: KalmanFilter
        """
        return self._set(outputSystemMatrices=True)

    def setCalculateSlidingLikelihood(self):
        """
        Optionally calculate a sliding likelihood across consecutive measurements

        Default is false

        :return: KalmanFilter
        """
        return self._set(calculateSlidingLikelihood=True)

    def setSlidingLikelihoodWindow(self, value):
        """
        Set the param for number of consecutive measurements to include in the total likelihood calculation

        Default is 1

        :param value: Integer
        :return: KalmanFilter
        """
        return self._set(slidingLikelihoodWindow=value) \
            .setCalculateSlidingLikelihood()